ethcore/src/tx_filter.rs

```diff
@@ -60,36 +63,50 @@ impl TransactionFilter {
 	/// Check if transaction is allowed at given block.
 	pub fn transaction_allowed<C: BlockInfo + CallContract>(&self, parent_hash: &H256, transaction: &SignedTransaction, client: &C) -> bool {
 		let mut cache = self.permission_cache.lock();
-
-		let tx_type = match transaction.action {
-			Action::Create => tx_permissions::CREATE,
+		let (tx_type, to) = match transaction.action {
+			Action::Create => (tx_permissions::CREATE, Address::new()),
 			Action::Call(address) => if client.code_hash(&address, BlockId::Hash(*parent_hash)).map_or(false, |c| c != KECCAK_EMPTY) {
-				tx_permissions::CALL
-			} else {
-				tx_permissions::BASIC
-			}
+					(tx_permissions::CALL, address)
+				} else {
+					(tx_permissions::BASIC, address)
+				}
 		};
 
 		let sender = transaction.sender();
+		let value = transaction.value;
 		let key = (*parent_hash, sender);
 
 		if let Some(permissions) = cache.get_mut(&key) {
 			return *permissions & tx_type != 0;
 		}
 
 		let contract_address = self.contract_address;
-		let permissions = self.contract.functions()
+
+		// Check permissions in smart contracts
```

Comment:
Can we mention here (or somewhere, where it's appropriate), that ABI of the contract with version=2 is used? Ideally is to add explicit check of the version in the code and force the next person, who will modify this code, to rely on contract's version. But perhaps at least some comment will be also helpful. PS. May be the fallback to the deprecated contract in case, when version is not retrieved, also will be looking more logically